File: strands-py/src/strands/vended_tools/mcp_router/mcp_router.py
"""Agent-callable MCP router tool.

Provides :func:`make_mcp_router` (a factory bound to a developer-set server allowlist).

The tool exposes five commands — ``connect``, ``list_connections``, ``list_tools``,
``call_tool``, ``disconnect`` — letting an agent open named connections to MCP servers,
inspect active connections, discover their tools, invoke them, and close the connections.
Each server is configured with a :class:`~strands.tools.mcp.MCPServerConfig`; all fields
are forwarded to :class:`~strands.tools.mcp.MCPClient`. Connections are isolated per agent.
"""

from __future__ import annotations

import asyncio
import logging
import threading
import weakref
from typing import TYPE_CHECKING, Any, Literal, cast
from uuid import uuid4

from ...tools.decorator import tool
from ...tools.mcp.mcp_agent_tool import MCPAgentTool
from ...tools.mcp.mcp_client import MCPClient, MCPServerConfig
from ...tools.mcp.mcp_types import MCPToolResult
from ...types.tools import ToolContext, ToolSpec
from .types import MCP_ROUTER_DESCRIPTION

if TYPE_CHECKING:
    from ...tools.decorator import DecoratedFunctionTool

logger = logging.getLogger(__name__)

_DEFAULT_MAX_CONNECTIONS = 10


class MCPRouterToolError(RuntimeError):
    """Raised when an mcp_router tool operation fails."""


def make_mcp_router(
    *,
    name: str = "mcp_router",
    description: str | None = None,
    servers: dict[str, MCPServerConfig],
    max_connections: int = _DEFAULT_MAX_CONNECTIONS,
) -> DecoratedFunctionTool:
    """Create an agent-callable MCP router tool bound to a developer-set server allowlist.

    MCP connections are scoped per agent and persist across invocations. They are closed when
    the model calls ``disconnect`` explicitly or when the agent is garbage collected.
    Connections remain open otherwise.

    Args:
        name: Tool name shown to the model.
        description: Tool description shown to the model. Defaults to a description
            that includes the permitted server names.
        servers: Allowlisted servers keyed by name. The model identifies servers to connect to by name;
            the config is forwarded to :class:`~strands.tools.mcp.MCPClient`. Must not be empty.
        max_connections: Maximum simultaneous open connections per agent. Defaults to ``10``.

    Returns:
        A decorated tool that manages MCP connections.

    Raises:
        ValueError: If ``servers`` is empty or ``max_connections`` is not positive.
    """
    if not servers:
        raise ValueError("`servers` must not be empty; the mcp_router tool requires at least one server")
    if max_connections < 1:
        raise ValueError("`max_connections` must be at least 1")

    if description is None:
        permitted = ", ".join(f"'{s}'" for s in sorted(servers))
        description = f"{MCP_ROUTER_DESCRIPTION} Permitted server names: {permitted}."

    # Per-agent connections with WeakKeyDictionary so agents can be garbage collected.
    connections_map: weakref.WeakKeyDictionary[Any, dict[str, MCPClient]] = weakref.WeakKeyDictionary()

    @tool(name=name, description=description, context="tool_context")
    async def mcp_router_tool(
        command: Literal["connect", "list_connections", "list_tools", "call_tool", "disconnect"],
        tool_context: ToolContext,
        connection_id: str | None = None,
        server_name: str | None = None,
        tool_name: str | None = None,
        arguments: dict[str, Any] | None = None,
    ) -> list[ToolSpec] | MCPToolResult | str:
        """Manage runtime MCP client connections.

        Args:
            command: The operation to perform: ``connect``, ``list_connections``,
                ``list_tools``, ``call_tool``, ``disconnect``.
            tool_context: Injected by the framework. Not user-facing.
            server_name: Server name to connect to, required for ``connect``.
            connection_id: A descriptive name for this connection, required for all commands
                except ``list_connections``. Must be unique per agent. Reusing an active id
                is rejected.
            tool_name: Tool name to invoke, required for ``call_tool``.
            arguments: Arguments to pass to the invoked tool, for ``call_tool``.

        Raises:
            MCPRouterToolError: If a required argument is missing, the server is not on
                the allowlist, the connection cap is reached, no matching connection exists,
                or the connection fails to start.
        """
        agent = tool_context.agent

        if command == "list_connections":
            return ", ".join(sorted(connections_map.get(agent, {})))

        if not connection_id:
            raise MCPRouterToolError("`connection_id` is required for all commands except 'list_connections'")

        if command == "connect":
            if not server_name:
                raise MCPRouterToolError("`server_name` is required for command='connect'")
            return await _handle_connect(connections_map, agent, servers, server_name, connection_id, max_connections)

        connections = connections_map.get(agent, {})
        client = connections.get(connection_id)
        if client is None:
            raise MCPRouterToolError(f"No active connection for id {connection_id!r}")

        if command == "list_tools":
            return await asyncio.to_thread(_handle_list_tools, client)

        if command == "call_tool":
            if not tool_name:
                raise MCPRouterToolError("`tool_name` is required for command='call_tool'")
            return await client.call_tool_async(
                tool_use_id=str(uuid4()),
                name=tool_name,
                arguments=arguments,
                cancel_signal=tool_context.cancel_signal,
            )

        if command == "disconnect":
            return await _handle_disconnect(connections, connection_id)

        raise MCPRouterToolError(f"Unknown command: {command}")

    return mcp_router_tool


# ---- Internals ----------------------------------------------------------------


def _stop_client(client: MCPClient) -> None:
    try:
        client.stop(None, None, None)
    except Exception:
        logger.debug("failed to stop MCP client", exc_info=True)


def _stop_clients_in_background(connections: dict[str, MCPClient]) -> None:
    threads = []
    for connection_id, client in list(connections.items()):
        logger.debug(
            "connection_id=<%s> | closing MCP connection during garbage collection",
            connection_id,
        )
        thread = threading.Thread(target=_stop_client, args=(client,), daemon=True)
        thread.start()
        threads.append(thread)
    for thread in threads:
        # Small timeout so a hanging stop does not block the garbage collector indefinitely.
        thread.join(timeout=1.0)


async def _handle_connect(
    connections_map: weakref.WeakKeyDictionary[Any, dict[str, MCPClient]],
    agent: Any,
    servers: dict[str, MCPServerConfig],
    server_name: str,
    connection_id: str,
    max_connections: int,
) -> str:
    if server_name not in servers:
        permitted = ", ".join(sorted(servers))
        raise MCPRouterToolError(f"Server {server_name!r} is not on the MCP server allowlist: {permitted}")

    config = cast(dict[str, Any], servers[server_name])
    loaded = MCPClient.load_servers({server_name: config})
    if not loaded:
        raise MCPRouterToolError(f"Server {server_name!r} failed to initialise; check the server config")
    client = loaded[0]

    try:
        # start() blocks until the MCP background thread signals ready — run it off the event loop.
        await asyncio.to_thread(client.start)
    except BaseException:
        _stop_client(client)
        raise

    # Cap and duplicate checks after all awaits to avoid race conditions.
    connections = connections_map.get(agent)
    if connections is None:
        connections = {}
        connections_map[agent] = connections
        # Stop all open connections if the agent is garbage collected without calling disconnect.
        weakref.finalize(agent, _stop_clients_in_background, connections)

    if len(connections) >= max_connections:
        _stop_client(client)
        active_ids = ", ".join(sorted(connections))
        raise MCPRouterToolError(
            f"Connection limit of {max_connections} reached. "
            f"Disconnect one of the active connections before opening a new one. "
            f"Active connection_ids: {active_ids}"
        )

    if connection_id in connections:
        _stop_client(client)
        raise MCPRouterToolError(
            f"Connection {connection_id!r} already exists. Disconnect it first or use a different connection_id."
        )

    connections[connection_id] = client

    logger.debug("connection_id=<%s>, server_name=<%s> | opened MCP connection", connection_id, server_name)
    return f"Successfully connected to {server_name} as {connection_id!r}"


def _handle_list_tools(client: MCPClient) -> list[ToolSpec]:
    all_tools: list[MCPAgentTool] = []
    pagination_token: str | None = None
    while True:
        page = client.list_tools_sync(pagination_token)
        all_tools.extend(page)
        pagination_token = page.pagination_token
        if pagination_token is None:
            break
    # Get mcp_tool.name instead of tool_name (may be prefixed) so call_tool works verbatim.
    return [{**t.tool_spec, "name": t.mcp_tool.name} for t in all_tools]


async def _handle_disconnect(connections: dict[str, MCPClient], connection_id: str) -> str:
    client = connections.pop(connection_id, None)
    if client is not None:
        await asyncio.to_thread(_stop_client, client)
    return "Successfully disconnected"

File: strands-ts/src/vended-tools/mcp-router/mcp-router.ts
import { tool } from '../../tools/tool-factory.js'
import { z } from 'zod'
import type { InvokableTool, ToolContext } from '../../tools/tool.js'
import type { ToolSpec } from '../../tools/types.js'
import type { JSONValue } from '../../types/json.js'
import type { ToolResultContent } from '../../types/messages.js'
import { McpClient, type McpServerConfig } from '../../mcp/index.js'
import { McpTool } from '../../tools/mcp-tool.js'
import { logger } from '../../logging/logger.js'
import { MCP_ROUTER_DESCRIPTION, McpRouterToolError } from './types.js'
import type { McpRouterInput } from './types.js'

const DEFAULT_MAX_CONNECTIONS = 10

/**
 * Zod schema for mcp_router input validation.
 */
const mcpRouterInputSchema = z.object({
  command: z
    .enum(['connect', 'list_connections', 'list_tools', 'call_tool', 'disconnect'])
    .describe("The operation to perform: 'connect', 'list_connections', 'list_tools', 'call_tool', 'disconnect'."),
  connection_id: z
    .string()
    .optional()
    .describe(
      'A descriptive name for this connection. Required for all commands except list_connections. ' +
        'Must be unique per agent. Reusing an active id is rejected.'
    ),
  server_name: z.string().optional().describe("Server name to connect to. Required for 'connect'."),
  tool_name: z.string().optional().describe("Tool name to invoke. Required for 'call_tool'."),
  arguments: z
    .record(z.string(), z.unknown())
    .optional()
    .describe("Arguments to pass to the invoked tool. Used with 'call_tool'."),
})

export interface MakeMcpRouterOptions {
  name?: string
  description?: string
  servers: Record<string, McpServerConfig>
  maxConnections?: number
}

/**
 * Creates an MCP router tool bound to a developer-set server allowlist.
 *
 * Connections are scoped per agent and persist across invocations. They are closed
 * when the model calls `disconnect` or when the agent is garbage collected.
 *
 * @param options - Configuration options.
 * @returns A tool that manages MCP connections.
 * @throws Error if `servers` is empty or `maxConnections` is not positive.
 *
 * @example
 * ```typescript
 * const mcpRouter = makeMcpRouter({ servers: { 'my-api': { url: 'https://mcp.example.com/mcp' } } })
 * const agent = new Agent({ tools: [mcpRouter] })
 * ```
 */
export function makeMcpRouter(options: MakeMcpRouterOptions): InvokableTool<McpRouterInput, JSONValue> {
  const { servers, maxConnections = DEFAULT_MAX_CONNECTIONS } = options

  if (Object.keys(servers).length === 0) {
    throw new Error('`servers` must not be empty; the mcp_router tool requires at least one server')
  }
  if (maxConnections < 1) {
    throw new Error('`maxConnections` must be at least 1')
  }

  const description =
    options.description ??
    `${MCP_ROUTER_DESCRIPTION} Permitted server names: ${Object.keys(servers)
      .sort()
      .map((s) => `'${s}'`)
      .join(', ')}.`

  // Per-agent connections with WeakMap so agents can be garbage collected.
  const connectionsMap = new WeakMap<object, Map<string, McpClient>>()

  // Cleans up open connections when an agent is garbage collected without calling disconnect.
  const registry = new FinalizationRegistry<Map<string, McpClient>>((connections) => {
    stopClientsInBackground(connections)
  })

  return tool({
    name: options.name ?? 'mcp_router',
    description,
    inputSchema: mcpRouterInputSchema,
    callback: async (input, context) => {
      if (!context) {
        throw new McpRouterToolError('mcp_router requires an agent context')
      }
      const { agent } = context

      const { command, connection_id: connectionId, server_name: serverName, tool_name: toolName } = input
      const args = (input.arguments ?? {}) as JSONValue

      if (command === 'list_connections') {
        const connections = connectionsMap.get(agent)
        const ids = connections ? [...connections.keys()] : []
        return ids.sort().join(', ')
      }

      if (!connectionId) {
        throw new McpRouterToolError("`connection_id` is required for all commands except 'list_connections'")
      }

      if (command === 'connect') {
        if (!serverName) {
          throw new McpRouterToolError("`server_name` is required for command='connect'")
        }
        return await handleConnect(connectionsMap, registry, agent, servers, serverName, connectionId, maxConnections)
      }

      const connections = connectionsMap.get(agent)
      const client = connections?.get(connectionId)
      if (!client) {
        throw new McpRouterToolError(`No active connection for id ${JSON.stringify(connectionId)}`)
      }

      if (command === 'list_tools') {
        return (await handleListTools(client)) as unknown as JSONValue
      }

      if (command === 'call_tool') {
        if (!toolName) {
          throw new McpRouterToolError("`tool_name` is required for command='call_tool'")
        }
        return await handleCallTool(client, toolName, args, context)
      }

      if (command === 'disconnect') {
        return await handleDisconnect(connections!, connectionId)
      }

      throw new McpRouterToolError(`Unknown command: ${command as string}`)
    },
  })
}

// ---- Internals ----------------------------------------------------------------

async function stopClient(client: McpClient): Promise<void> {
  try {
    await client.disconnect()
  } catch {
    logger.debug('failed to stop MCP client')
  }
}

function stopClientsInBackground(connections: Map<string, McpClient>): void {
  for (const [connectionId, client] of connections) {
    logger.debug(`connection_id=<${connectionId}> | closing MCP connection during garbage collection`)
    void stopClient(client)
  }
}

async function handleConnect(
  connectionsMap: WeakMap<object, Map<string, McpClient>>,
  registry: FinalizationRegistry<Map<string, McpClient>>,
  agent: object,
  servers: Record<string, McpServerConfig>,
  serverName: string,
  connectionId: string,
  maxConnections: number
): Promise<string> {
  if (!Object.hasOwn(servers, serverName)) {
    const permitted = Object.keys(servers).sort().join(', ')
    throw new McpRouterToolError(
      `Server ${JSON.stringify(serverName)} is not on the MCP server allowlist: ${permitted}`
    )
  }

  const serverConfig = servers[serverName]!
  const loaded = await McpClient.loadServers({ [serverName]: serverConfig })
  if (loaded.length === 0) {
    throw new McpRouterToolError(`Server ${JSON.stringify(serverName)} failed to initialise; check the server config`)
  }
  const client = loaded[0]!

  try {
    await client.connect()
  } catch (error) {
    await stopClient(client)
    throw error
  }

  // McpClient.connect() swallows failures when continueOnError is set in the server config.
  if (client.connectionState === 'failed') {
    await stopClient(client)
    throw new McpRouterToolError(`Server ${JSON.stringify(serverName)} failed to connect; check the server config`)
  }

  // Cap and duplicate checks after all awaits to avoid race conditions.
  let connections = connectionsMap.get(agent)
  if (!connections) {
    connections = new Map()
    connectionsMap.set(agent, connections)
    // Stop all open connections if the agent is garbage collected without calling disconnect.
    registry.register(agent, connections)
  }

  if (connections.size >= maxConnections) {
    await stopClient(client)
    const activeIds = [...connections.keys()].sort().join(', ')
    throw new McpRouterToolError(
      `Connection limit of ${maxConnections} reached. ` +
        `Disconnect one of the active connections before opening a new one. ` +
        `Active connection_ids: ${activeIds}`
    )
  }

  if (connections.has(connectionId)) {
    await stopClient(client)
    throw new McpRouterToolError(
      `Connection ${JSON.stringify(connectionId)} already exists. Disconnect it first or use a different connection_id.`
    )
  }

  connections.set(connectionId, client)

  logger.debug(`connection_id=<${connectionId}>, server_name=<${serverName}> | opened MCP connection`)
  return `Successfully connected to ${serverName} as ${JSON.stringify(connectionId)}`
}

async function handleListTools(client: McpClient): Promise<ToolSpec[]> {
  const tools = await client.listTools({ prefix: '' })
  // Get tool.name (server-side name without prefix) so call_tool works verbatim.
  return tools.map((t) => ({ ...t.toolSpec, name: t.name }))
}

async function handleCallTool(
  client: McpClient,
  toolName: string,
  args: JSONValue,
  context: ToolContext
): Promise<JSONValue> {
  // Ad-hoc tool uses tool.name, which the model finds from an earlier listTools call.
  const tempTool = new McpTool({
    name: toolName,
    description: '',
    inputSchema: { type: 'object' },
    client,
  })
  // Using McpTool.stream instead of McpClient.callTool to map content to SDK blocks and errors
  const gen = tempTool.stream({
    ...context,
    toolUse: { ...context.toolUse, name: toolName, input: args },
  })
  let next = await gen.next()
  while (!next.done) next = await gen.next()
  const result = next.value
  if (result.status === 'error') {
    const text = result.content.map((c: ToolResultContent) => ('text' in c ? c.text : JSON.stringify(c))).join('\n')
    throw new McpRouterToolError(text || 'MCP tool returned an error')
  }
  return result.content as unknown as JSONValue
}

async function handleDisconnect(connections: Map<string, McpClient>, connectionId: string): Promise<string> {
  const client = connections.get(connectionId)
  connections.delete(connectionId)
  if (client) {
    await stopClient(client)
  }
  return 'Successfully disconnected'
}
